File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.py
#가치가 최대인 것을 우선으로 하고 
# 무게가 되지마자 출력

n, max_weight = map(int, input().split())
items = []

for _ in range(n):
    weight, value = map(int, input().split())
    items.append((weight, value))

dp = [0] * (max_weight + 1)

for weight, value in items:
    for w in range(max_weight, weight - 1, -1):
        dp[w] = max(dp[w], dp[w - weight] + value)

print(max(dp))

   

File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.js
const fs = require('fs');
const filePath = process.platform === 'linux' ? '/dev/stdin' : './input.txt';
const input = fs.readFileSync(filePath).toString().trim();

function solution(input) {
  const parseInput = (input || '').toString().split(`\n`);
  const [N, K] = parseInput[0].split(' ').map(Number);
  const nums = parseInput.slice(1).map(a => a.split(' ').map(Number));

  // 가치 안넘으면 계속 갱신

  const dp = new Array(K + 1).fill(0);

  for (let [weight, value] of nums) {
    for (let i = K; i >= weight; i--) {
      dp[i] = Math.max(dp[i], dp[i - weight] + value);
    }
  }

  return dp[dp.length - 1];
}

console.log(solution(input));
